### DISCUSSION TITLE
Webassembly publish to linux-arm64

### DISCUSSION THREAD

**Discussion by ashah7** [unoplatform/uno#18097] [General]
I have created a new project using Visual Studio 2022 v17.11.2
I'm trying to publish my app in Webassembly but getting the below during the publish screen, I dont see any options to pick a platform or have a platform specific version.  Going ahead and publishing it fails anyway (seems to default to win-x64

![image](https://github.com/user-attachments/assets/6862fd00-9b0a-4326-af7b-3245287d4967)

```
Severity    Code    Description    Project    File    Line    Suppression State
Error        Assets file '....\obj\project.assets.json' doesn't have a target for 'net8.0-browserwasm/win-x64'. Ensure that restore has run and that you have included 'net8.0-browserwasm' in the TargetFrameworks for your project. You may also need to include 'win-x64' in your project's RuntimeIdentifiers.    Menu   

```
If I try to publish using command line
`dotnet publish --configuration Release --no-self-contained --framework "net8.0-browserwasm" --runtime "linux-arm64"`

It seems to publish but the resulting application does not run on my arm64 device (Raspberry pi)

```

 Menu-> G:\Repos\Menu\Menu\bin\Release\net8.0-browserwasm\linux-arm64\Menu.dll
 Menu-> G:\Repos\Menu\Menu\bin\Release\net8.0-browserwasm\linux-arm64\publish\

```

in the publish folder I see a `Menu` file making this executable and running it fails 
I've coped the .dll file into a folder on my pi together with everything from publish but they all fail

I'm I doing something obvious wrong?
Compiling and running as a desktop app runs fine on the pi

```

pi@1000000064fe2d06:~/Menu $ dotnet MenuWeb.dll
Unhandled exception. System.TypeInitializationException: TypeInitialization_Type, <Module>
 ---> System.IO.FileNotFoundException: Could not load file or assembly 'Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.

IO_FileName_Name, Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null
   at __UnoUseOpenSansInitializer.Initialize()
   at .cctor()
   Exception_EndOfInnerExceptionStack
Aborted
pi@1000000064fe2d06:~/Menu $ chmod +x MenuWeb.dll
pi@1000000064fe2d06:~/Menu $ ./MenuWeb
You must install .NET to run this application.

App: /home/pi/Menu/Menu
Architecture: arm64
App host version: 8.0.8
.NET location: Not found

Learn more:
https://aka.ms/dotnet/app-launch-failed

Download the .NET runtime:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=arm64&rid=linux-arm64&os=debian.12&apphost_version=8.0.8
pi@1000000064fe2d06:~/Menu $ dotnet Menu
Could not execute because the specified command or file was not found.
Possible reasons for this include:
  * You misspelled a built-in dotnet command.
  * You intended to execute a .NET program, but dotnet-Menu does not exist.
  * You intended to run a global tool, but a dotnet-prefixed executable with this name could not be found on the PATH.

```

**Comment by ashah7** [comment]
I've made some progress here, unlike blazor or razor apps, uno doesn't come with its own webserver like kestrel.  So it has to be manually done using nginx or apache or dotnet serve

Anyhow I managed to copy the files from the publish locations  to my linux machine 
```

 Menu-> G:\Repos\Menu\Menu\bin\Release\net8.0-browserwasm\linux-arm64\Menu.dll
 Menu-> F:\www\Menu
```
I also copied the Menu.dll over not sure in which directory this goes but I stuck it in /home/pi/Menu (above the wwwroot)

and ran `dotnet serve -d /home/pi/Menu/wwwroot/ -a any -v --port 5000`

when I cannot to port 5000 I see my application popup all fine and dandy

However I do run a local Linux command that seems to error with the below "platform not supported message".

Has anyone has a similar experience?  I suspect perhaps I've copied the .dll in the wrong place or its still not compiled properly for arm64

A similar example would be be running "`uname -a`" and displaying that on  the webpage


```

invoke-js.ts:176 fail: Uno.UI.Dispatching.NativeDispatcher[0]
      NativeDispatcher unhandled exception
System.PlatformNotSupportedException: Process_PlatformNotSupported
   at System.Diagnostics.Process..ctor()
   at Menu.Models.LinuxCmdHelper.Bash(String cmd)
   at Menu.Presentation.WifiConfigViewModel.GetNetworksAction()
   at Menu.Presentation.WifiConfigViewModel.<GetNetworks>b__13_0()
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at Uno.UI.Dispatching.NativeDispatcherSynchronizationContext.<>c__DisplayClass3_0.<Post>b__0()
   at Uno.UI.Dispatching.NativeDispatcher.DispatchItems() 
```

**Reply by jeromelaban** [reply]
Thanks for the report. The last error message generally means that the command you're executing is running the wasm app as if it were a desktop app, which cannot work. Could you provide the actual command line that you use to "do run a local linux command"?

Otherwise, for the publication of apps, it's best to use [the command line experience](https://platform.uno/docs/articles/uno-publishing-webassembly.html).

**Reply by ashah7** [reply]
so the command can be anything platform related say I wanted to display the kernel version on a web page e.g. `uname -a`

To get this running I use a helper class like the below, and run the command using something like the below in my code behind
`
"uname -a".Bash();`

However this doesnt work it gives me the above platform error

Now, I know I this does work in a traditional webapp using a ajax query.
Question is how can I get this working here with Webassembly?  

This understandably works fine on my desktop version of the app, but Uno is supposed to be cross platform so why doesnt it work in Webassembly?   I guess my understanding of cross platform isn't as ambitious as expected

It would also be good to hear some thoughts on how the experts would accomplish a similar task

```
public static class LinuxCmdHelper
{

    public static string Bash(this string cmd)
    {
        var escapedArgs = cmd.Replace("\"", "\\\"");

        var process = new Process()
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = "/bin/bash",
                Arguments = $"-c \"{escapedArgs}\"",
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true,

            }
        };
        process.Start();
        string result = process.StandardOutput.ReadToEnd();
        process.WaitForExit();

        return result.Trim() + ";" + process.ExitCode;
    }

```

**Reply by jeromelaban** [reply]
I see, thanks for the details.

> This understandably works fine on my desktop version of the app, but Uno is supposed to be cross platform so why doesnt it work in Webassembly? I guess my understanding of cross platform isn't as ambitious as expected

Cross platform does not mean that all APIs can work everywhere. On WebAssembly, specifically, `Process` cannot work because it's running inside the browser VM, which has no such concept. If you want to get the environment information, you'll need to get it differently, if possible, through JS interop. iOS, Android and Catalyst all have their own set of limitations based on what's supported in their environment.
